rest/stargate.go

```diff
@@ -0,0 +1,129 @@
+package rest
+
+import (
+	"fmt"
+	"net/url"
+	"strings"
+
+	"github.com/cosmos/cosmos-sdk/types/query"
+	"github.com/cosmos/cosmos-sdk/types/tx"
+	"github.com/gin-gonic/gin"
+	"github.com/jackc/pgx/v4/pgxpool"
+	"github.com/likecoin/likecoin-chain-tx-indexer/db"
+	"github.com/likecoin/likecoin-chain-tx-indexer/logger"
+)
+
+func getOffset(query url.Values) (uint64, error) {
+	return getUint(query, "pagination.offset")
+}
+
+func isOrderByDesc(query url.Values) (bool, error) {
```

Comment:
Here you are hijacking a `bool` to represent an ordering, which I do not prefer since it is arbitrary and unfriendly to the reader (is `true` representing `ASC` or `DESC`?).

I prefer having something like an enum, e.g.
```Go
const (
    ORDER_ASC = "ASC"
    ORDER_DESC = "DESC"
)

func parseQueryOrder(query url.Values) (string, error) {
    // ...
    switch blablabla {
    case blablabla:
        return ORDER_ASC, nil
    case blablabla:
        return ORDER_DESC, nil
    default:
        return "", fmt.Errorf(blablabla)
    }
}
```

(In this case I used `string` as the underlying type, sometimes we may want even more complicated types, it depends on the coding context. I use `string` since eventually we want to plug it into SQL statement, so it would be convenient and easy to read)

Reply:
I see, I thought I should prevent directly parse param to SQL statements `ASC` & `DESC` not until the Model layer (which is db.go) at the beginning. So I used a `bool`, but it's indeed arbitrary.

Reply:
On the type level, you don't actually care about the underlying value (binary level / wrapped type / whatever), just like you don't care `'a'` is actually `97`, `0x61` or `0b01100001` inside the memory, all you know is that you can use `'a'` to refer to this character and do some operations (e.g. comparison, `toUpper` which outputs `'A'`).
In the case of indexer query, you only refer to those values in variable names and only need comparison (in switch case or if ... else if ... else), so the underlying values are not important, whether they are `true / false`, `"ASC" / "DESC"` or `1 / 2` or `1337 / 0xc0de`.
What is important in this level is that you have assigned them a name (in variables or constants, or enum if the language supports that), and the names are readable for others to understand. So you may actually use `true / false` but just wrap them into meaningful names and types.
You also would like to consider if you need extension for this enum-like type, e.g. if you use `bool` but then you have a 3rd option in the future, then you will need to modify all return types referring this enum-like type.
So usually we will use numbers as enum, since it is extensible, and is faster than string when doing comparison.
Using a string here is somewhat a hack, since in this case we want to eventually convert it into a string, so using string as underlying type will be a convenient choice.